File: _archive/utils/csv_handler.py
# Created by DINKIssTyle on 2026.
# Copyright (C) 2026 DINKI'ssTyle. All rights reserved.

"""CSV 가져오기/내보내기 핸들러"""

import csv
import io
from typing import List, Tuple, Optional


def detect_delimiter(text: str) -> str:
    """텍스트에서 구분자 자동 감지 (탭 또는 쉼표)"""
    first_line = text.split('\n')[0] if text else ""
    tab_count = first_line.count('\t')
    comma_count = first_line.count(',')
    return '\t' if tab_count > comma_count else ','


def import_csv(file_path: str) -> Tuple[List[str], List[List[str]]]:
    """CSV 파일을 읽어서 헤더와 데이터 반환"""
    with open(file_path, 'r', encoding='utf-8-sig') as f:
        content = f.read()

    delimiter = detect_delimiter(content)
    reader = csv.reader(io.StringIO(content), delimiter=delimiter)
    rows = list(reader)

    if not rows:
        return [], []

    headers = rows[0]
    data = rows[1:]
    return headers, data


def import_csv_text(text: str) -> Tuple[List[str], List[List[str]]]:
    """CSV 텍스트를 파싱하여 헤더와 데이터 반환"""
    delimiter = detect_delimiter(text)
    reader = csv.reader(io.StringIO(text), delimiter=delimiter)
    rows = list(reader)

    if not rows:
        return [], []

    headers = rows[0]
    data = rows[1:]
    return headers, data


def export_csv(file_path: str, headers: List[str], data: List[List[str]],
               delimiter: str = ','):
    """데이터를 CSV 파일로 내보내기"""
    with open(file_path, 'w', encoding='utf-8-sig', newline='') as f:
        writer = csv.writer(f, delimiter=delimiter)
        writer.writerow(headers)
        for row in data:
            writer.writerow(row)


def export_csv_text(headers: List[str], data: List[List[str]],
                    delimiter: str = ',') -> str:
    """데이터를 CSV 텍스트로 변환"""
    output = io.StringIO()
    writer = csv.writer(output, delimiter=delimiter)
    writer.writerow(headers)
    for row in data:
        writer.writerow(row)
    return output.getvalue()

File: utils/csv_handler.go
// Created by DINKIssTyle on 2026.
// Copyright (C) 2026 DINKI'ssTyle. All rights reserved.

package utils

import (
	"encoding/csv"
	"strings"
	"unicode/utf8"
)

// DetectDelimiter detects if the CSV is comma or tab separated
func DetectDelimiter(text string) rune {
	lines := strings.SplitN(text, "\n", 2)
	if len(lines) == 0 {
		return ','
	}
	firstLine := lines[0]
	tabCount := strings.Count(firstLine, "\t")
	commaCount := strings.Count(firstLine, ",")
	if tabCount > commaCount {
		return '\t'
	}
	return ','
}

// ImportCSVText parses CSV/TSV text and returns headers and data rows
func ImportCSVText(text string) ([]string, [][]string, error) {
	delimiter := DetectDelimiter(text)
	r := csv.NewReader(strings.NewReader(text))
	r.Comma = delimiter
	r.LazyQuotes = true

	rows, err := r.ReadAll()
	if err != nil {
		return nil, nil, err
	}

	if len(rows) == 0 {
		return []string{}, [][]string{}, nil
	}

	headers := rows[0]
	data := rows[1:]
	return headers, data, nil
}

// ExportCSVText converts headers and data to CSV/TSV string
func ExportCSVText(headers []string, data [][]string, delimiter rune) (string, error) {
	var sb strings.Builder
	// Add UTF-8 BOM for Excel compatibility (matching Python's utf-8-sig)
	sb.Write([]byte{0xEF, 0xBB, 0xBF})

	w := csv.NewWriter(&sb)
	w.Comma = delimiter

	if err := w.Write(headers); err != nil {
		return "", err
	}
	for _, row := range data {
		if err := w.Write(row); err != nil {
			return "", err
		}
	}
	w.Flush()
	return sb.String(), nil
}

// EnsureValidUTF8 ensures the string is valid UTF-8
func EnsureValidUTF8(s string) string {
	if utf8.ValidString(s) {
		return s
	}
	v := make([]rune, 0, len(s))
	for i, r := range s {
		if r == utf8.RuneError {
			_, size := utf8.DecodeRuneInString(s[i:])
			if size == 1 {
				continue
			}
		}
		v = append(v, r)
	}
	return string(v)
}
